sdk/eventhub/azure-eventhubs/azure/eventhub/aio/eventprocessor/event_processor.py

```diff
@@ -205,96 +100,175 @@ def _context(self, events):
+         and asynchronously start receiving EventData from EventHub and processing events.
+
+        :return: None
+
+        """
+        log.info("EventProcessor %r is being started", self._id)
+        ownership_manager = OwnershipManager(self._eventhub_client, self._consumer_group_name, self._id,
+                                             self._partition_manager, self._ownership_timeout, self._partition_id)
+        if not self._running:
+            self._running = True
+            while self._running:
+                try:
+                    checkpoints = await ownership_manager.get_checkpoints() if self._partition_manager else None
+                    claimed_partition_ids = await ownership_manager.claim_ownership()
+                    if claimed_partition_ids:
+                        to_cancel_list = self._tasks.keys() - claimed_partition_ids
+                        self._create_tasks_for_claimed_ownership(claimed_partition_ids, checkpoints)
+                    else:
+                        log.info("EventProcessor %r hasn't claimed an ownership. It keeps claiming.", self._id)
+                        to_cancel_list = set(self._tasks.keys())
+                    self._cancel_tasks_for_partitions(to_cancel_list)
+                except Exception as err:  # pylint:disable=broad-except
+                    '''
+                    ownership_manager.get_checkpoints() and ownership_manager.claim_ownership() may raise exceptions
+                    when there are load balancing and/or checkpointing (partition_manager isn't None).
+                    They're swallowed here to retry every self._polling_interval seconds. Meanwhile this event processor
+                    won't lose the partitions it has claimed before.
+                    If it keeps failing, other EventProcessors will start to claim ownership of the partitions
+                    that this EventProcessor is working on. So two or multiple EventProcessors may be working
+                    on the same partition.
+                    '''  # pylint:disable=pointless-string-statement
+                    log.warning("An exception (%r) occurred during balancing and claiming ownership for "
+                                "eventhub %r consumer group %r. Retrying after %r seconds",
+                                err, self._eventhub_name, self._consumer_group_name, self._polling_interval)
+                await asyncio.sleep(self._polling_interval)
+
+    async def stop(self):
+        """Stop the EventProcessor.
+
+        The EventProcessor will stop receiving events from EventHubs and release the ownership of the partitions
+        it is working on.
+        Other running EventProcessor will take over these released partitions.
+
+        A stopped EventProcessor can be restarted by calling method `start` again.
+
+        :return: None
+
+        """
+        self._running = False
+        pids = list(self._tasks.keys())
```

Comment:
don't need to wrap with `list` if you're just iterating on it

Reply:
During the execution, self._tasks will remove items. The for ... loop will be corrupt if the looped collection is changed.